### PR TITLE
[MUGEN] Add Multimodal Transformer Decoder

### PR DESCRIPTION
**Pull request by langong347**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #205
* #213
* #204
* #176

Summary:
- Added the transformer decoder of GPT model following paper [Improving Language Understanding
by Generative Pre-Training](https://cdn.openai.com/research-covers/language-unsupervised/language_understanding_paper.pdf)
- Extended for cross-modality generation
- Refactored [`P2QAttentionStack`](https://github.com/mugen-org/MUGEN_baseline/blob/main/lib/models/gpt/attention.py#L15) from MUGEN research project

Test Plan:

```
$ $ python -m pytest --cov=torchmultimodal/models/ test/models/test_gpt.py -vv
================================================= test session starts ==================================================
platform darwin -- Python 3.8.13, pytest-7.1.2, pluggy-1.0.0 -- /Users/langong/local/miniconda3/envs/torchmm/bin/python
cachedir: .pytest_cache
rootdir: /Users/langong/gpt_attention, configfile: pyproject.toml
plugins: mock-3.8.2, cov-3.0.0
collected 14 items

test/models/test_gpt.py::TestMultimodalGPT::test_bad_input PASSED                                                [  7%]
test/models/test_gpt.py::TestMultimodalGPT::test_forward_in_modality PASSED                                      [ 14%]
test/models/test_gpt.py::TestMultimodalGPT::test_forward_out_modality PASSED                                     [ 21%]
test/models/test_gpt.py::TestMultimodalGPT::test_forward_two_modality PASSED                                     [ 28%]
test/models/test_gpt.py::TestMultimodalGPT::test_encode PASSED                                                   [ 35%]
test/models/test_gpt.py::TestMultimodalGPT::test_bad_pos_ids PASSED                                              [ 42%]
test/models/test_gpt.py::TestMultimodalGPT::test_optional_pos_ids PASSED                                         [ 50%]
test/models/test_gpt.py::TestTransformerDecoder::test_forward PASSED                                             [ 57%]
test/models/test_gpt.py::TestTransformerDecoder::test_forward_additional_output PASSED                           [ 64%]
test/models/test_gpt.py::TestTransformerDecoderLayer::test_forward PASSED                                        [ 71%]
test/models/test_gpt.py::TestTransformerDecoderLayer::test_forward_masked PASSED                                 [ 78%]
test/models/test_gpt.py::TestTransformerDecoderLayer::test_forward_additional_output PASSED                      [ 85%]
test/models/test_gpt.py::test_sigmoid_linear_unit PASSED                                                         [ 92%]
test/models/test_gpt.py::test_right_shift PASSED                                                                 [100%]

---------- coverage: platform darwin, python 3.8.13-final-0 ----------
Name                                                  Stmts   Miss  Cover
-------------------------------------------------------------------------
torchmultimodal/models/gpt.py                           123      0   100%
-------------------------------------------------------------------------
TOTAL                                                  1169   1046    11%


====================== 14 passed in 1.77s ==============
```

Differential Revision: [D38316990](https://our.internmc.facebook.com/intern/diff/D38316990)

### CODE REVIEW

File: torchmultimodal/models/gpt.py [hunk 1]
```diff
@@ -0,0 +1,418 @@
...
+    Attributes:
+        in_token_emb (nn.Module): input modality token embedding layer.
+        out_token_emb (nn.Module): output modality, i.e., the modality to be generated,
+            token embedding layer.
+        in_pos_emb (nn.Module): input modality position embedding layer.
+        out_pos_emb (nn.Module): output modality position embedding layer.
+        decoder (nn.Module): the transformer decoder (see ``torchmultimodal.models.gpt.TransformerDecoder``)
+
+    Args:
+        in_modality (Tensor, optional): Tensor of dimension ``(bs, d1, ..., dn, c)`` containing data for the
```
Comment (RdoubleA) [inline torchmultimodal/models/gpt.py:54]: should we keep the naming consistent and use `b` instead of `bs` for batch dim?

File: torchmultimodal/models/gpt.py [hunk 2]
```diff
@@ -0,0 +1,418 @@
...
+            x,
+            attn_mask,
+            head_mask,
+            use_cache,
+            causal,
+            return_attn_weights,
+            return_hidden_states,
+        )
+
+        return TransformerDecoderOutput(
```
Comment (RdoubleA) [inline torchmultimodal/models/gpt.py:138]: why not just return out?

File: torchmultimodal/models/gpt.py [hunk 3]
```diff
@@ -0,0 +1,418 @@
...
+            last_hidden_states=hidden_states,
+            hidden_states=all_hidden_states,
+            attention_weights=all_attentions,
+            past_key_values=all_past_key_values,
+        )
+
+
+class SiLU(nn.Module):
+    r"""Sigmoind Linear Unit
```
Comment (RdoubleA) [inline torchmultimodal/models/gpt.py:244]:
```suggestion
    r"""Sigmoid Linear Unit
```

File: torchmultimodal/models/gpt.py [hunk 4]
```diff
@@ -0,0 +1,418 @@
...
+        causal: bool = False,
+        return_attn_weights: bool = False,
+        return_hidden_states: bool = False,
+    ) -> TransformerDecoderOutput:
+        if (in_modality is None) and (out_modality is None):
+            raise ValueError(
+                "in_modality and out_modality sequences cannot be both empty"
+            )
+
+        # Since generation is based the previous data point (autoregressive) where
```
Comment (RdoubleA) [inline torchmultimodal/models/gpt.py:109]:
```suggestion
        # Since generation is based on the previous data point (autoregressive) where
```

File: torchmultimodal/models/gpt.py [hunk 5]
```diff
@@ -0,0 +1,418 @@
...
+            * self.sos
+        )  # (batch, 1, embedding_dim)
+        sos = sos.type_as(x)
+        # Shift one unit to the right along dim ``seq_len``
+        x = torch.cat([sos, x[:, :-1, :]], axis=1)  # (batch, seq_len, embedding_dim)
+        x = x.view(*x_shape)
+        return x
+
+
+def _get_clones(module: nn.Module, n: int) -> nn.Module:
```
Comment (RdoubleA) [inline torchmultimodal/models/gpt.py:417]: should we make this a common utility since MDETR also uses this?

Comment (langong347) [reply torchmultimodal/models/gpt.py:417]: I can move that to utils. @ebsmothers

File: torchmultimodal/models/gpt.py [hunk 6]
```diff
@@ -0,0 +1,405 @@
...
+
+        return TransformerDecoderOutput(
+            last_hidden_states=hidden_states,
+            hidden_states=all_hidden_states,
+            attention_weights=all_attentions,
+            past_key_values=all_past_key_values,
+        )
+
+
+class SiLU(nn.Module):
```
Comment (ebsmothers) [inline torchmultimodal/models/gpt.py:235]: We also have this in CLIP now so might be worth consolidating: https://github.com/facebookresearch/multimodal/blob/1c4fa802684a87a26ca203ef18bf3110ab10ff1a/torchmultimodal/modules/encoders/clip_text_encoder.py#L14

File: torchmultimodal/models/gpt.py [hunk 7]
```diff
@@ -0,0 +1,405 @@
...
+            head_mask,
+            use_cache,
+            causal,
+            return_attn_weights,
+            return_hidden_states,
+        )
+
+    def _encode(self, x: Tensor, pos_ids: Tensor, modality: str) -> Tensor:
+        token_emb = getattr(self, f"{modality}_token_emb")
+        pos_emb = getattr(self, f"{modality}_pos_emb")
```
Comment (ebsmothers) [inline torchmultimodal/models/gpt.py:139]: Why do we do things this way if we are just gonna use in/out_token/pos_emb in init anyways?

File: torchmultimodal/models/gpt.py [hunk 8]
```diff
@@ -0,0 +1,405 @@
...
+            x,
+            attention_mask=attn_mask,
+            head_mask=head_mask,
+            return_attn_weights=return_attn_weights,
+            use_cache=use_cache,
+            causal=causal,
+        )
+
+    @checkpoint_wrapper
+    def _mlp_block(self, x: Tensor) -> Tensor:
```
Comment (ebsmothers) [inline torchmultimodal/models/gpt.py:373]: Does this just get its own method for the purpose of checkpointing?

File: torchmultimodal/models/gpt.py [hunk 9]
```diff
@@ -0,0 +1,405 @@
...
+        return self.mlp(x)
+
+
+class RightShift(nn.Module):
```
Comment (ebsmothers) [inline torchmultimodal/models/gpt.py:377]: Where else is this used?

File: torchmultimodal/models/gpt.py [hunk 10]
```diff
@@ -0,0 +1,405 @@
...
+    def __init__(self, embedding_dim: int) -> None:
+        super().__init__()
+        self.embedding_dim = embedding_dim
+        self.sos = nn.Parameter(torch.FloatTensor(embedding_dim).normal_(std=0.02))
+
+    def forward(self, x: Tensor) -> Tensor:
+        x_shape = list(x.shape)
+        x = x.flatten(start_dim=1, end_dim=-2)  # (batch, seq_len, embedding_dim)
+        sos = (
+            torch.ones(x_shape[0], 1, self.embedding_dim, dtype=torch.float32).to(
```
Comment (ebsmothers) [inline torchmultimodal/models/gpt.py:395]: probably don't need dtype here if you are just casting to self.sos immediately afterwards

A separate point: Why not just do

`sos = self.sos.unsqueeze(0).unsqueeze(1).repeat(x_shape[0],1,1)`

(or something like that)?

File: torchmultimodal/models/gpt.py [hunk 11]
```diff
@@ -0,0 +1,405 @@
...
+    def _encode(self, x: Tensor, pos_ids: Tensor, modality: str) -> Tensor:
+        token_emb = getattr(self, f"{modality}_token_emb")
+        pos_emb = getattr(self, f"{modality}_pos_emb")
+        _, x = token_emb(x)  # (tokens, embedding per token)
+        x = x + pos_emb(pos_ids)
+
+        return x
+
+    def _norm_pos_ids(self, x: Tensor, pos_ids: Optional[Tensor] = None) -> Tensor:
+        b, seq_len, _ = x.shape
```
Comment (ebsmothers) [inline torchmultimodal/models/gpt.py:146]: Do we need to assert that x has the right number of dims?

File: test/models/test_gpt.py [hunk 12]
```diff
@@ -0,0 +1,311 @@
...
+
+
+def tuple_to_dict(t):
+    if not isinstance(t, tuple):
+        raise TypeError(f"Input must be of type tuple but got {type(t)}")
+
+    return {k: v for k, v in enumerate(t)}
+
+
+def assert_expected_wrapper(actual, expected):
```
Comment (ebsmothers) [inline test/models/test_gpt.py:32]: Worth putting in test_utils?

File: torchmultimodal/models/gpt.py (hunk 6 above)
Comment (langong347) [reply torchmultimodal/models/gpt.py:235]: I noticed that but not sure what our stance here since we are gradually shifting towards model centric components.
And it will be weird for `gpt.py` to take a dep on `clip.py` -- could be confusing to users.
Or we can create `torchmultimodal/modules/layers/activation.py` and move it over there?

File: torchmultimodal/models/gpt.py (hunk 8 above)
Comment (langong347) [reply torchmultimodal/models/gpt.py:373]: yes

File: torchmultimodal/models/gpt.py (hunk 9 above)
Comment (langong347) [reply torchmultimodal/models/gpt.py:377]: Not in this PR but later:
- This needs to be part of training as `sos` is a model parameter. 
- The trained `sos` will be the first input token during generation.

Context:
- MUGEN [inserts](https://github.com/mugen-org/MUGEN_baseline/blob/main/lib/models/gpt/attention.py#L84) `RightShift` in the middle of GPT after position embedding and before transformer decoder. 
- RightShift only gets triggered in the two modes I mentioned above and their implementation IMO is quite [hacky](https://github.com/mugen-org/MUGEN_baseline/blob/main/lib/models/gpt/attention.py#L318).

My plan:
- Trigger `RightShift` when `self.training` is `True` such that I don't have to break apart GPT into position embedding + RightShift + TransformerDecoder.
- During generation load the trained `sos` from the weight file to initialize the sequence to be generated

File: torchmultimodal/models/gpt.py (hunk 6 above)
Comment (ebsmothers) [reply torchmultimodal/models/gpt.py:226]: Yeah agree that we should not take a dep on CLIP here. I like the option of adding activation.py but fine to leave as-is for now if you prefer

File: torchmultimodal/models/gpt.py (hunk 7 above)
Comment (ebsmothers) [reply torchmultimodal/models/gpt.py]: Looks good after the latest updates 🙂

File: torchmultimodal/models/gpt.py (hunk 11 above)
Comment (ebsmothers) [reply torchmultimodal/models/gpt.py:137]: Thanks for clarifying. I think the renaming makes sense

**Review by ebsmothers — APPROVED** [review]
Looks good to me. Up to you if you want to create the separate activation.py file we discussed, but either way is fine with me

File: torchmultimodal/models/gpt.py (hunk 6 above)
Comment (langong347) [reply torchmultimodal/models/gpt.py:226]: Ok. Created a task to move to activation.py as a follow-up.

**Comment by langong347** [comment]
@langong347 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38316990).
